src/arch/riscv/isa.cc

```diff
@@ -779,6 +780,31 @@ ISA::setMiscReg(RegIndex idx, RegVal val)
                 setMiscRegNoEffect(idx, new_val);
             }
             break;
+          case MISCREG_SENVCFG:
+            {
+                // panic on write to bitfields that aren't implemented in gem5
+                SENVCFG panic_mask = 0;
+                panic_mask.pmm = 3;
+
+                SENVCFG wpri_mask = 0;
+                wpri_mask.wpri_1 = ~wpri_mask.wpri_1;
+                wpri_mask.wpri_2 = ~wpri_mask.wpri_2;
+                wpri_mask.wpri_3 = ~wpri_mask.wpri_3;
+
+                if ((panic_mask & val) != 0) {
+                    panic("Tried to write to an unimplemented bitfield in the "
+                    "senvcfg CSR!\nThe attempted write was:\n %" PRIu64 "\n",
+                    val);
+                }
+
+                if ((wpri_mask & val) != 0) {
+                    warn("Ignoring write to WPRI bit(s) in senvcfg CSR.\n"
+                    "The attempted write was:\n %" PRIu64 "\n", val);
+                } else {
+                    setMiscRegNoEffect(idx, val);
```

Comment:
Can you simply call `setMiscReg`? I think we want the effect.

Reply:
I've left this call as setMiscRegNoEffect, as this part of the code is already in setMiscReg. It's just used as a setter function here with the code above being the effect. I think that's how the other registers also use setMiscRegNoEffect.

I've also changed the implementation so `val` excludes the WPRI bits.